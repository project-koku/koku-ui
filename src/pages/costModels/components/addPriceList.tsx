import {
  ActionGroup,
  Button,
  ButtonVariant,
  Form,
  Stack,
  StackItem,
  Text,
  TextContent,
  TextVariants,
  Title,
  TitleSizes,
} from '@patternfly/react-core';
import { MetricHash } from 'api/metrics';
import { intl as defaultIntl } from 'components/i18n';
import messages from 'locales/messages';
import {
  canSubmit as isReadyForSubmit,
  RateForm,
  RateFormData,
  useRateData,
} from 'pages/costModels/components/rateForm';
import { CostModelContext } from 'pages/costModels/createCostModelWizard/context';
import React from 'react';
import { injectIntl, WrappedComponentProps } from 'react-intl';

interface AddPriceListOwnProps {
  metricsHash: MetricHash;
  submitRate: (data: RateFormData) => void;
  cancel: () => void;
}

type AddPriceListProps = AddPriceListOwnProps & WrappedComponentProps;

const AddPriceList: React.FunctionComponent<AddPriceListProps> = ({
  intl = defaultIntl, // Default required for testing
  submitRate,
  cancel,
  metricsHash,
}) => {
  const { tiers } = React.useContext(CostModelContext);
  const rateFormData = useRateData(metricsHash, undefined, tiers);
  const canSubmit = React.useMemo(() => isReadyForSubmit(rateFormData), [rateFormData.errors, rateFormData.rateKind]);
  return (
    <Stack hasGutter>
      <StackItem>
        <Title headingLevel="h2" size={TitleSizes.xl}>
          {intl.formatMessage(messages.CostModelsWizardCreatePriceList)}
        </Title>
      </StackItem>
      <StackItem>
        <TextContent>
          <Text component={TextVariants.h6}>{intl.formatMessage(messages.CostModelsWizardPriceListMetric)}</Text>
        </TextContent>
      </StackItem>
      <StackItem>
        <Form>
          <RateForm metricsHash={metricsHash} rateFormData={rateFormData} />
        </Form>
      </StackItem>
      <StackItem>
        <ActionGroup>
          <Button variant={ButtonVariant.primary} isDisabled={!canSubmit} onClick={() => submitRate(rateFormData)}>
            {intl.formatMessage(messages.CreateRate)}
          </Button>
          <Button variant={ButtonVariant.link} onClick={cancel}>
            {intl.formatMessage(messages.Cancel)}
          </Button>
        </ActionGroup>
      </StackItem>
    </Stack>
  );
};

export default injectIntl(AddPriceList);
