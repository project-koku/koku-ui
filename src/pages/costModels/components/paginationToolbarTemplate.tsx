import { Pagination, PaginationProps, Toolbar, ToolbarContent, ToolbarItem } from '@patternfly/react-core';
import React from 'react';

interface PaginationToolbarTemplateProps extends PaginationProps {
  id?: string;
}

export const PaginationToolbarTemplate: React.SFC<PaginationToolbarTemplateProps> = ({
  id,
  itemCount,
  perPage,
  page,
  variant,
  onPerPageSelect,
  onSetPage,
}) => {
  return (
    <Toolbar id={id}>
      <ToolbarContent>
        <ToolbarItem variant="pagination">
          <Pagination
            variant={variant}
            itemCount={itemCount}
            perPage={perPage}
            page={page}
            onSetPage={onSetPage}
            onPerPageSelect={onPerPageSelect}
          />
        </ToolbarItem>
      </ToolbarContent>
    </Toolbar>
  );
};

export default PaginationToolbarTemplate;
