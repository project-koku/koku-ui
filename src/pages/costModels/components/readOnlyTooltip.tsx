import { Tooltip } from '@patternfly/react-core';
import React from 'react';

interface ReadOnlyTooltipBase {
  tooltip?: string;
  children: JSX.Element;
  isDisabled: boolean;
}

export const ReadOnlyTooltip: React.FunctionComponent<ReadOnlyTooltipBase> = ({
  children,
  tooltip = 'You have read only permissions',
  isDisabled,
}) => {
  return isDisabled ? (
    <Tooltip isContentLeftAligned content={<div>{tooltip}</div>}>
      <div aria-label="Read only">{children}</div>
    </Tooltip>
  ) : (
    children
  );
};
