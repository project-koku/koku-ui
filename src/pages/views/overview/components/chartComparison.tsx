import { Dropdown, DropdownItem, DropdownToggle } from '@patternfly/react-core';
import React from 'react';

interface ChartComparisonOwnProps {
  currentItem?: string;
  onItemClicked(value: string);
  options?: {
    label: string;
    value: string;
  }[];
}

interface ChartComparisonState {
  isChartComparisonOpen: boolean;
}

type ChartComparisonProps = ChartComparisonOwnProps;

class ChartComparisonBase extends React.Component<ChartComparisonProps> {
  protected defaultState: ChartComparisonState = {
    isChartComparisonOpen: false,
  };
  public state: ChartComparisonState = { ...this.defaultState };

  private getDropDownItems = () => {
    const { options } = this.props;

    return options.map(option => (
      <DropdownItem component="button" key={option.value} onClick={() => this.handleClick(option.value)}>
        {option.label}
      </DropdownItem>
    ));
  };

  private getCurrentLabel = () => {
    const { currentItem, options } = this.props;

    let label = '';
    for (const option of options) {
      if (currentItem === option.value) {
        label = option.label;
        break;
      }
    }
    return label;
  };

  public handleClick = value => {
    const { onItemClicked } = this.props;
    if (onItemClicked) {
      onItemClicked(value);
    }
  };

  private handleSelect = () => {
    this.setState({
      isChartComparisonOpen: !this.state.isChartComparisonOpen,
    });
  };

  private handleToggle = isChartComparisonOpen => {
    this.setState({
      isChartComparisonOpen,
    });
  };

  public render() {
    const { isChartComparisonOpen } = this.state;
    const dropdownItems = this.getDropDownItems();

    return (
      <Dropdown
        onSelect={this.handleSelect}
        toggle={<DropdownToggle onToggle={this.handleToggle}>{this.getCurrentLabel()}</DropdownToggle>}
        isOpen={isChartComparisonOpen}
        dropdownItems={dropdownItems}
      />
    );
  }
}

const ChartComparison = ChartComparisonBase;

export { ChartComparison };
