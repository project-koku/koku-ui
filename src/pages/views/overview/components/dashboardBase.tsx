import { Grid, GridItem } from '@patternfly/react-core';
import React from 'react';
import { WrappedComponentProps } from 'react-intl';

type DashboardOwnProps = WrappedComponentProps;

interface DashboardStateProps {
  DashboardWidget: any;
  widgets: number[];
}

interface DashboardDispatchProps {
  selectWidgets?: () => void;
}

type DashboardProps = DashboardOwnProps & DashboardStateProps & DashboardDispatchProps;

const DashboardBase: React.SFC<DashboardProps> = ({ DashboardWidget, selectWidgets, widgets }) => (
  <div>
    <Grid hasGutter>
      {widgets.map(widgetId => {
        const widget = selectWidgets[widgetId];
        return widget.details.showHorizontal ? (
          <GridItem sm={12} key={widgetId}>
            <DashboardWidget widgetId={widgetId} />
          </GridItem>
        ) : (
          <GridItem lg={12} xl={6} xl2={4} key={widgetId}>
            <DashboardWidget widgetId={widgetId} />
          </GridItem>
        );
      })}
    </Grid>
  </div>
);

export { DashboardBase };
