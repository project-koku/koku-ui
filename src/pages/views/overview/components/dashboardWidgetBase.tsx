import { Tab, Tabs, TabTitleText } from '@patternfly/react-core';
import { Forecast } from 'api/forecasts/forecast';
import { getQuery } from 'api/queries/awsQuery';
import { Report } from 'api/reports/report';
import {
  ChartType,
  ComputedReportItemType,
  transformForecast,
  transformForecastCone,
  transformReport,
} from 'components/charts/common/chartDatumUtils';
import {
  ReportSummary,
  ReportSummaryAlt,
  ReportSummaryCost,
  ReportSummaryDailyCost,
  ReportSummaryDailyTrend,
  ReportSummaryDetails,
  ReportSummaryItem,
  ReportSummaryItems,
  ReportSummaryTrend,
  ReportSummaryUsage,
} from 'components/reports/reportSummary';
import { format, getMonth, startOfMonth } from 'date-fns';
import { cloneDeep } from 'lodash';
import React from 'react';
import { WithTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';
import { DashboardChartType, DashboardWidget } from 'store/dashboard/common/dashboardCommon';
import { formatValue, unitLookupKey } from 'utils/formatValue';

import { ChartComparison } from './chartComparison';
import { chartStyles, styles } from './dashboardWidget.styles';

// eslint-disable-next-line no-shadow
const enum Comparison {
  cumulative = 'cumulative',
  daily = 'daily',
}

interface DashboardWidgetOwnProps {
  chartAltHeight?: number;
  containerAltHeight?: number;
  getIdKeyForTab: <T extends DashboardWidget<any>>(tab: T) => string;
  widgetId: number;
}

interface DashboardWidgetStateProps extends DashboardWidget<any> {
  currentQuery: string;
  currentReport: Report;
  currentReportFetchStatus: number;
  forecast?: Forecast;
  previousQuery: string;
  previousReport: Report;
  tabsQuery: string;
  tabsReport: Report;
  tabsReportFetchStatus: number;
}

interface DashboardWidgetDispatchProps {
  fetchForecasts: (widgetId) => void;
  fetchReports: (widgetId) => void;
  updateTab: (id, availableTabs) => void;
}

type DashboardWidgetProps = DashboardWidgetOwnProps &
  DashboardWidgetStateProps &
  DashboardWidgetDispatchProps &
  WithTranslation;

class DashboardWidgetBase extends React.Component<DashboardWidgetProps> {
  public state = {
    activeTabKey: 0,
    currentComparison: Comparison.cumulative,
  };

  public componentDidMount() {
    const { availableTabs, fetchForecasts, fetchReports, id, trend, updateTab, widgetId } = this.props;

    if (availableTabs) {
      updateTab(id, availableTabs[0]);
    }
    if (fetchReports) {
      fetchReports(widgetId);
    }
    if (trend.computedForecastItem !== undefined) {
      fetchForecasts(widgetId);
    }
  }

  private buildDetailsLink = <T extends DashboardWidget<any>>(tab: T) => {
    const { details, getIdKeyForTab } = this.props;
    const currentTab = getIdKeyForTab(tab);
    return `${details.viewAllPath}?${getQuery({
      group_by: {
        [currentTab]: '*',
      },
      order_by: { cost: 'desc' },
    })}`;
  };

  private getChart = (containerHeight: number, height: number, adjustContainerHeight: boolean = false) => {
    const { chartType, trend } = this.props;
    if (chartType === DashboardChartType.dailyTrend) {
      return this.getDailyTrendChart(
        containerHeight,
        height,
        adjustContainerHeight,
        trend.showInfrastructureLabel,
        trend.showSupplementaryLabel
      );
    } else if (chartType === DashboardChartType.dailyCost) {
      return this.getDailyCostChart(containerHeight, height, adjustContainerHeight);
    } else if (chartType === DashboardChartType.trend) {
      return this.getTrendChart(
        containerHeight,
        height,
        adjustContainerHeight,
        trend.showInfrastructureLabel,
        trend.showSupplementaryLabel
      );
    } else if (chartType === DashboardChartType.usage) {
      return this.getUsageChart(height, adjustContainerHeight);
    } else {
      return null;
    }
  };

  private getChartComparison = () => {
    const { t, trend } = this.props;
    const { currentComparison } = this.state;

    const units = this.getUnits();
    const cumulativeTitle = t(trend.titleKey, { units: t(`units.${units}`) });
    const dailyTitle = t(trend.dailyTitleKey, { units: t(`units.${units}`) });

    const options = [
      { label: dailyTitle, value: Comparison.daily },
      { label: cumulativeTitle, value: Comparison.cumulative },
    ];

    return (
      <ChartComparison
        currentItem={currentComparison || options[0].value}
        onItemClicked={this.handleComparisonClick}
        options={options}
      />
    );
  };

  // This chart displays cumulative and daily cost compared to infrastructure cost
  private getDailyCostChart = (containerHeight: number, height: number, adjustContainerHeight: boolean = false) => {
    const { currentReport, previousReport, trend } = this.props;
    const { currentComparison } = this.state;

    const computedReportItem = trend.computedReportItem; // cost, supplementary cost, etc.
    const computedReportItemValue = trend.computedReportItemValue; // infrastructure usage cost

    const daily = currentComparison === Comparison.daily;
    const type = daily ? ChartType.daily : trend.type;

    // Infrastructure data
    const currentInfrastructureData = transformReport(
      currentReport,
      type,
      'date',
      'infrastructure',
      computedReportItemValue
    );
    const previousInfrastructureData = transformReport(
      previousReport,
      type,
      'date',
      'infrastructure',
      computedReportItemValue
    );

    // Cost data
    const currentCostData = transformReport(currentReport, type, 'date', computedReportItem, computedReportItemValue);
    const previousCostData = transformReport(previousReport, type, 'date', computedReportItem, computedReportItemValue);

    // Forecast data
    const forecastData = this.getForecastData(currentReport, trend.computedForecastItem);
    const forecastInfrastructureData = this.getForecastData(currentReport, trend.computedForecastInfrastructureItem);

    const ReportSummaryComponent = daily ? ReportSummaryDailyCost : ReportSummaryCost;
    return (
      <>
        <div style={styles.comparison}>{this.getChartComparison()}</div>
        <ReportSummaryComponent
          adjustContainerHeight={adjustContainerHeight}
          containerHeight={containerHeight}
          currentCostData={currentCostData}
          currentInfrastructureCostData={currentInfrastructureData}
          forecastConeData={forecastData.forecastConeData}
          forecastData={forecastData.forecastData}
          forecastInfrastructureConeData={forecastInfrastructureData.forecastConeData}
          forecastInfrastructureData={forecastInfrastructureData.forecastData}
          formatDatumValue={formatValue}
          formatDatumOptions={trend.formatOptions}
          height={height}
          previousCostData={previousCostData}
          previousInfrastructureCostData={previousInfrastructureData}
          showForecast={trend.computedForecastItem !== undefined}
        />
      </>
    );
  };

  // This chart displays cumulative and daily cost
  private getDailyTrendChart = (
    containerHeight: number,
    height: number,
    adjustContainerHeight: boolean = false,
    showInfrastructureLabel: boolean = false,
    showSupplementaryLabel: boolean = false
  ) => {
    const { currentReport, details, previousReport, trend } = this.props;
    const { currentComparison } = this.state;

    const units = this.getUnits();
    const computedReportItem = trend.computedReportItem; // cost, supplementary cost, etc.
    const computedReportItemValue = trend.computedReportItemValue; // infrastructure usage cost

    const daily = currentComparison === Comparison.daily;
    const type = daily ? ChartType.daily : trend.type;

    // Cost data
    const currentData = transformReport(currentReport, type, 'date', computedReportItem, computedReportItemValue);
    const previousData = transformReport(previousReport, type, 'date', computedReportItem, computedReportItemValue);

    // Forecast data
    const { forecastData, forecastConeData } = this.getForecastData(currentReport, trend.computedForecastItem);

    const ReportSummaryComponent = daily ? ReportSummaryDailyTrend : ReportSummaryTrend;
    return (
      <>
        <div style={styles.comparison}>{this.getChartComparison()}</div>
        <ReportSummaryComponent
          adjustContainerHeight={adjustContainerHeight}
          containerHeight={containerHeight}
          currentData={currentData}
          forecastData={forecastData}
          forecastConeData={forecastConeData}
          formatDatumValue={formatValue}
          formatDatumOptions={trend.formatOptions}
          height={height}
          previousData={previousData}
          showForecast={trend.computedForecastItem !== undefined}
          showInfrastructureLabel={showInfrastructureLabel}
          showSupplementaryLabel={showSupplementaryLabel}
          showUsageLegendLabel={details.showUsageLegendLabel}
          units={units}
        />
      </>
    );
  };

  private getForecastData = (report: Report, computedForecastItem: string = 'cost') => {
    const { forecast, trend } = this.props;
    const { currentComparison } = this.state;

    // Todo: Add cumulative / daily prop
    const daily = currentComparison === Comparison.daily;
    const type = daily ? ChartType.daily : trend.type;

    let forecastData;
    let forecastConeData;

    if (computedForecastItem) {
      const newForecast = cloneDeep(forecast);
      if (newForecast) {
        newForecast.data = [];
      }
      if (forecast && report && report.data) {
        const total =
          report.meta && report.meta.total && report.meta.total[computedForecastItem]
            ? report.meta.total[computedForecastItem].total.value
            : 0;

        // Find last currentData date with values
        const reportedValues = report.data.filter(val => val.values.length);
        const lastReported = reportedValues[reportedValues.length - 1]
          ? reportedValues[reportedValues.length - 1].date
          : undefined;

        // Remove overlapping forecast dates, if any
        if (forecast && forecast.data && forecast.data.length > 0) {
          const lastReportedDate = new Date(lastReported);
          const lastReportedMonth = lastReportedDate.getMonth() + 1;
          for (const item of forecast.data) {
            const forecastDate = new Date(item.date);
            const forecastMonth = forecastDate.getMonth() + 1;

            // Ensure month match. AWS forecast may begin with "2020-12-04", but ends on "2021-01-01"
            if (forecastDate > lastReportedDate && lastReportedMonth === forecastMonth) {
              newForecast.data.push(item);
            }
          }

          // For cumulative data, forecast values should begin at last reported total with zero confidence values
          if (type === ChartType.rolling) {
            const firstReported =
              forecast.data[0].values && forecast.data[0].values.length > 0
                ? forecast.data[0].values[0].date
                : undefined;

            const date = this.getNumberOfDays(lastReported, firstReported) === 1 ? lastReported : firstReported;

            newForecast.data.unshift({
              date,
              values: [
                {
                  date,
                  cost: {
                    confidence_max: {
                      value: 0,
                    },
                    confidence_min: {
                      value: 0,
                    },
                    total: {
                      value: total,
                      units: 'USD',
                    },
                  },
                  infrastructure: {
                    confidence_max: {
                      value: 0,
                    },
                    confidence_min: {
                      value: 0,
                    },
                    total: {
                      value: total,
                      units: 'USD',
                    },
                  },
                  supplementary: {
                    confidence_max: {
                      value: 0,
                    },
                    confidence_min: {
                      value: 0,
                    },
                    total: {
                      value: total,
                      units: 'USD',
                    },
                  },
                },
              ],
            });
          }
        }
      }
      forecastData = transformForecast(newForecast, type, computedForecastItem);
      forecastConeData = transformForecastCone(newForecast, type, computedForecastItem);
    }
    return { forecastData, forecastConeData };
  };

  private getNumberOfDays = (start: string, end: string) => {
    const date1 = new Date(start);
    const date2 = new Date(end);

    // One day in milliseconds
    const oneDay = 1000 * 60 * 60 * 24;

    // Calculating the time difference between two dates
    const diffInTime = date2.getTime() - date1.getTime();

    // Calculating the no. of days between two dates
    const diffInDays = Math.round(diffInTime / oneDay);

    return diffInDays;
  };

  // This chart displays cumulative cost only
  private getTrendChart = (
    containerHeight: number,
    height: number,
    adjustContainerHeight: boolean = false,
    showInfrastructureLabel: boolean = false,
    showSupplementaryLabel: boolean = false
  ) => {
    const { currentReport, details, previousReport, t, trend } = this.props;

    const units = this.getUnits();
    const title = t(trend.titleKey, { units: t(`units.${units}`) });
    const computedReportItem = trend.computedReportItem; // cost, supplementary cost, etc.
    const computedReportItemValue = trend.computedReportItemValue; // infrastructure usage cost

    // Cost data
    const currentData = transformReport(currentReport, trend.type, 'date', computedReportItem, computedReportItemValue);
    const previousData = transformReport(
      previousReport,
      trend.type,
      'date',
      computedReportItem,
      computedReportItemValue
    );

    // Forecast data
    const { forecastData, forecastConeData } = this.getForecastData(currentReport, trend.computedForecastItem);

    return (
      <ReportSummaryTrend
        adjustContainerHeight={adjustContainerHeight}
        containerHeight={containerHeight}
        currentData={currentData}
        forecastData={forecastData}
        forecastConeData={forecastConeData}
        formatDatumValue={formatValue}
        formatDatumOptions={trend.formatOptions}
        height={height}
        previousData={previousData}
        showForecast={trend.computedForecastItem !== undefined}
        showInfrastructureLabel={showInfrastructureLabel}
        showSupplementaryLabel={showSupplementaryLabel}
        showUsageLegendLabel={details.showUsageLegendLabel}
        title={title}
        units={units}
      />
    );
  };

  // This chart displays usage and requests
  private getUsageChart = (height: number, adjustContainerHeight: boolean = false) => {
    const { currentReport, previousReport, t, trend } = this.props;

    const units = this.getUnits();
    const title = t(trend.titleKey, { units: t(`units.${units}`) });

    // Request data
    const currentRequestData = transformReport(currentReport, trend.type, 'date', 'request');
    const previousRequestData = transformReport(previousReport, trend.type, 'date', 'request');

    // Usage data
    const currentUsageData = transformReport(currentReport, trend.type, 'date', 'usage');
    const previousUsageData = transformReport(previousReport, trend.type, 'date', 'usage');

    return (
      <ReportSummaryUsage
        adjustContainerHeight={adjustContainerHeight}
        containerHeight={chartStyles.containerUsageHeight}
        currentRequestData={currentRequestData}
        currentUsageData={currentUsageData}
        formatDatumValue={formatValue}
        formatDatumOptions={trend.formatOptions}
        height={height}
        previousRequestData={previousRequestData}
        previousUsageData={previousUsageData}
        title={title}
      />
    );
  };

  private getDetails = () => {
    const { chartType, currentReport, details, reportType, trend } = this.props;
    const computedReportItem = trend.computedReportItem || 'cost';
    const computedReportItemValue = trend.computedReportItemValue || 'total';
    const units = this.getUnits();

    return (
      <ReportSummaryDetails
        chartType={chartType}
        computedReportItem={computedReportItem}
        computedReportItemValue={computedReportItemValue}
        costLabel={this.getDetailsLabel(details.costKey, units)}
        formatOptions={details.formatOptions}
        formatValue={formatValue}
        report={currentReport}
        reportType={reportType}
        requestLabel={this.getDetailsLabel(details.requestKey, units)}
        showTooltip={details.showTooltip}
        showUnits={details.showUnits}
        showUsageFirst={details.showUsageFirst}
        units={details.units}
        usageFormatOptions={details.usageFormatOptions}
        usageLabel={this.getDetailsLabel(details.usageKey, units)}
      />
    );
  };

  private getDetailsLabel = (key: string, units: string) => {
    const { t } = this.props;
    return key ? t(key, { units: t(`units.${units}`) }) : undefined;
  };

  private getDetailsLink = () => {
    const { currentTab, details } = this.props;

    if (details.viewAllPath) {
      return <Link to={this.buildDetailsLink(currentTab)}>{this.getDetailsLinkTitle(currentTab)}</Link>;
    }
    return null;
  };

  private getDetailsLinkTitle = <T extends DashboardWidget<any>>(tab: T) => {
    const { getIdKeyForTab, t } = this.props;
    const key = getIdKeyForTab(tab) || '';

    return t('group_by.all', { groupBy: key });
  };

  private getHorizontalLayout = () => {
    const {
      containerAltHeight = chartStyles.containerAltHeight,
      chartAltHeight = chartStyles.chartAltHeight,
      currentReportFetchStatus,
      details,
    } = this.props;

    return (
      <ReportSummaryAlt
        detailsLink={this.getDetailsLink()}
        status={currentReportFetchStatus}
        tabs={this.getTabs()}
        title={this.getTitle()}
      >
        {this.getDetails()}
        {this.getChart(containerAltHeight, chartAltHeight, details.adjustContainerHeight)}
      </ReportSummaryAlt>
    );
  };

  private getTab = <T extends DashboardWidget<any>>(tab: T, index: number) => {
    const { getIdKeyForTab, tabsReport, tabsReportFetchStatus } = this.props;
    const currentTab: any = getIdKeyForTab(tab);

    return (
      <Tab
        eventKey={index}
        key={`${getIdKeyForTab(tab)}-tab`}
        title={<TabTitleText>{this.getTabTitle(tab)}</TabTitleText>}
      >
        <div style={styles.tabItems}>
          <ReportSummaryItems
            idKey={currentTab}
            key={`${currentTab}-items`}
            report={tabsReport}
            status={tabsReportFetchStatus}
          >
            {({ items }) => items.map(reportItem => this.getTabItem(tab, reportItem))}
          </ReportSummaryItems>
        </div>
      </Tab>
    );
  };

  private getTabItem = <T extends DashboardWidget<any>>(tab: T, reportItem) => {
    const { availableTabs, details, getIdKeyForTab, tabsReport, topItems, trend } = this.props;
    const { activeTabKey } = this.state;

    const currentTab = getIdKeyForTab(tab);
    const activeTab = getIdKeyForTab(availableTabs[activeTabKey]);
    const computedReportItem = trend.computedReportItem || 'cost';
    const computedReportItemValue = trend.computedReportItemValue || 'total';

    let totalValue;
    const hasTotal = tabsReport && tabsReport.meta && tabsReport.meta.total;
    if (computedReportItem === ComputedReportItemType.usage) {
      if (hasTotal && tabsReport.meta.total.usage) {
        totalValue = tabsReport.meta.total.usage.value;
      }
    } else {
      if (
        hasTotal &&
        tabsReport.meta.total[computedReportItem] &&
        tabsReport.meta.total[computedReportItem][computedReportItemValue]
      ) {
        totalValue = tabsReport.meta.total[computedReportItem][computedReportItemValue].value;
      }
    }

    if (activeTab === currentTab) {
      return (
        <ReportSummaryItem
          key={`${reportItem.id}-item`}
          formatOptions={topItems.formatOptions}
          formatValue={formatValue}
          label={reportItem.label ? reportItem.label.toString() : ''}
          totalValue={totalValue}
          units={details.units ? details.units : this.getUnits()}
          value={reportItem[computedReportItem][computedReportItemValue].value}
        />
      );
    } else {
      return null;
    }
  };

  private getTabs = () => {
    const { availableTabs } = this.props;
    return (
      <Tabs isFilled activeKey={this.state.activeTabKey} onSelect={this.handleTabClick}>
        {availableTabs.map((tab, index) => this.getTab(tab, index))}
      </Tabs>
    );
  };

  private getTabTitle = <T extends DashboardWidget<any>>(tab: T) => {
    const { getIdKeyForTab, t } = this.props;
    const key = getIdKeyForTab(tab) || '';

    return t('group_by.top', { groupBy: key });
  };

  private getTitle = () => {
    const { t, titleKey } = this.props;

    const today = new Date();
    const month = getMonth(today);
    const endDate = format(today, 'Do');
    const startDate = format(startOfMonth(today), 'Do');

    return t(titleKey, { endDate, month, startDate });
  };

  private getUnits = () => {
    const { currentReport, details, trend } = this.props;
    const computedReportItem = trend.computedReportItem || 'cost';
    const computedReportItemValue = trend.computedReportItemValue || 'total';

    if (details.units) {
      return details.units;
    }

    let units;
    const hasTotal = currentReport && currentReport.meta && currentReport.meta.total;
    if (computedReportItem === ComputedReportItemType.usage) {
      const hasUsage = hasTotal && currentReport.meta.total.usage;
      units = hasUsage ? unitLookupKey(currentReport.meta.total.usage.units) : '';
    } else {
      const hasCost =
        hasTotal &&
        currentReport.meta.total[computedReportItem] &&
        currentReport.meta.total[computedReportItem][computedReportItemValue];
      units = hasCost ? unitLookupKey(currentReport.meta.total[computedReportItem][computedReportItemValue].units) : '';
    }
    return units;
  };

  private getVerticalLayout = () => {
    const { availableTabs, currentReportFetchStatus, details } = this.props;

    return (
      <ReportSummary detailsLink={this.getDetailsLink()} status={currentReportFetchStatus} title={this.getTitle()}>
        {this.getDetails()}
        {this.getChart(chartStyles.containerTrendHeight, chartStyles.chartHeight, details.adjustContainerHeight)}
        {Boolean(availableTabs) && <div style={styles.tabs}>{this.getTabs()}</div>}
      </ReportSummary>
    );
  };

  private handleComparisonClick = (value: string) => {
    this.setState({ currentComparison: value });
  };

  private handleTabClick = (event, tabIndex) => {
    const { availableTabs, id, updateTab } = this.props;
    const tab = availableTabs[tabIndex];

    updateTab(id, tab);
    this.setState({
      activeTabKey: tabIndex,
    });
  };

  public render() {
    const { details } = this.props;
    return details.showHorizontal ? this.getHorizontalLayout() : this.getVerticalLayout();
  }
}

export { DashboardWidgetBase, DashboardWidgetProps, DashboardWidgetOwnProps, DashboardWidgetStateProps };
