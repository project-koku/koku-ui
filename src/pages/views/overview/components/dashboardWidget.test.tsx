jest.mock('date-fns').mock('date-fns/format');

import { MessageDescriptor } from '@formatjs/intl/src/types';
import { ChartType } from 'components/charts/common/chartDatumUtils';
import { createIntlEnv } from 'components/i18n/localeEnv';
import { format, getDate, getMonth, startOfMonth } from 'date-fns';
import { shallow } from 'enzyme';
import messages from 'locales/messages';
import { DashboardWidgetBase, DashboardWidgetProps } from 'pages/views/overview/components/dashboardWidgetBase';
import React from 'react';
import { defineMessages } from 'react-intl';
import { FetchStatus } from 'store/common';
import { mockDate } from 'testUtils';

const tmessages = defineMessages({
  TestTitle: {
    id: 'TestTitle',
    description: 'test title',
    defaultMessage: 'test title',
  },
  TestTrendTitle: {
    id: 'TestTrendTitle',
    description: 'test trend title',
    defaultMessage: 'test trend title',
  },
});

const props: DashboardWidgetProps = {
  widgetId: 1,
  id: 1,
  t: jest.fn(v => v),
  current: { data: [] },
  previous: { data: [] },
  tabs: { data: [] },
  fetchReports: jest.fn(),
  updateTab: jest.fn(),
  titleKey: tmessages.TestTitle,
  trend: {
    type: ChartType.rolling,
    titleKey: tmessages.TestTrendTitle,
    formatOptions: {},
  },
  status: FetchStatus.none,
  currentQuery: '',
  previousQuery: '',
  tabsQuery: '',
  details: {
    breakdownDescKeyRange: 'detail description range',
    breakdownDescKeySingle: 'detail description single',
    formatOptions: {},
  },
  topItems: { formatOptions: {} },
} as any;

const getDateMock = getDate as jest.Mock;
const formatMock = format as jest.Mock;
const startOfMonthMock = startOfMonth as jest.Mock;
const getMonthMock = getMonth as jest.Mock;

beforeEach(() => {
  mockDate();
  getDateMock.mockReturnValue(1);
  formatMock.mockReturnValue('formatted date');
  startOfMonthMock.mockReturnValue(1);
  getMonthMock.mockReturnValue(1);
});

test('reports are fetched on mount', () => {
  shallow(<DashboardWidgetBase {...props} />);
  expect(props.fetchReports).toBeCalledWith(props.widgetId);
});

test('title is translated', () => {
  shallow(<DashboardWidgetBase {...props} />);
  expect(getTranslateCallForKey(props.titleKey)).toMatchSnapshot();
});

test('subtitle is translated with single date', () => {
  shallow(<DashboardWidgetBase {...props} />);
  expect(getTranslateCallForKeyWithCnt(messages.DashBoardWidgetSubTitle, 1)).toMatchSnapshot();
});

test('subtitle is translated with date range', () => {
  getDateMock.mockReturnValueOnce(2);
  shallow(<DashboardWidgetBase {...props} />);
  expect(getTranslateCallForKeyWithCnt(messages.DashBoardWidgetSubTitle, 2)).toMatchSnapshot();
});

test('trend title is translated', () => {
  shallow(<DashboardWidgetBase {...props} />);
  expect(getTranslateCallForKey(props.trend.titleKey)).toMatchSnapshot();
});

function getTranslateCallForKeyWithCnt(key: MessageDescriptor, cnt: number) {
  const intl = createIntlEnv();
  const startDate = 1;
  const endDate = 15;

  return intl.formatMessage(key, { count: cnt, startDate, endDate, month: 'January' });
}

function getTranslateCallForKey(key: MessageDescriptor) {
  const intl = createIntlEnv();
  return intl.formatMessage(key);
}
