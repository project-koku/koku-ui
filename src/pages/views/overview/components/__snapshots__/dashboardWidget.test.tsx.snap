// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subtitle is translated with date range 1`] = `"EN 1-15 January"`;

exports[`subtitle is translated with single date 1`] = `"EN 1 January"`;

exports[`title is translated 1`] = `"test title"`;

exports[`trend title is translated 1`] = `"test trend title"`;
